File: global_custom/custom/js/sales_invoice.js
frappe.ui.form.on('Sales Invoice Item', {
	items_add: function(frm){
    console.log('Add row clicked')
		frm.set_df_property('si_itemwise_rate_details', 'hidden', 1);
	},
  item_code: function (frm, cdt, cdn) {
    var d = locals[cdt][cdn]
    const set_fields = ['rate','date','customer','sales_invoice'];
    if(!d.item_code){
      cur_frm.set_df_property('si_itemwise_rate_details', 'hidden', 1);
    }
    if(d.item_code){
      cur_frm.set_df_property('si_itemwise_rate_details', 'hidden', 0);
      frappe.call({
        method: "global_custom.custom.python.sales_invoice.fetch_rate_details",
        args: {
          item_code: d.item_code
        },
        freeze: true,
        callback: function (r) {
            if(r.message) {
                cur_frm.set_value('si_itemwise_rate_details', []);
                $.each(r.message, function(i, d) {
                    var row = cur_frm.add_child('si_itemwise_rate_details');
                    for (let key in d) {
                        if (d[key] && in_list(set_fields, key)) {
                            row[key] = d[key];
                        }
                    }
                });
            }
            refresh_field('si_itemwise_rate_details');
        }
      })
    }
  }
});

{% include 'erpnext/selling/sales_common.js' %};
frappe.provide("erpnext.accounts");


erpnext.accounts.SalesInvoiceController = erpnext.selling.SellingController.extend({
delivery_note_btn: function() {
  var me = this;
  this.$delivery_note_btn = this.frm.add_custom_button(__('Delivery Note'),
    function() {
      erpnext.utils.map_current_doc({
        method: "erpnext.stock.doctype.delivery_note.delivery_note.make_sales_invoice",
        source_doctype: "Delivery Note",
        target: me.frm,
        date_field: "posting_date",
        setters: {
          customer: me.frm.doc.customer || undefined,
          posting_date: undefined
        },
        get_query: function() {
          var filters = {
            docstatus: 1,
            company: me.frm.doc.company,
            is_return: 0
          };
          if(me.frm.doc.customer) filters["customer"] = me.frm.doc.customer;
          return {
            query: "erpnext.controllers.queries.get_delivery_notes_to_be_billed",
            filters: filters
          };
        }
      });
    }, __("Get Items From"));
}
})

File: global_custom/custom/python/sales_invoice.py
from frappe.custom.doctype.custom_field.custom_field import create_custom_fields
import frappe
from frappe import _
def unlink(doc,action):
    for row in doc.items:
        frappe.db.set_value("Sales Invoice Item",{"parent":doc.name},"delivery_note",None)

def make_custom_fields(update=True):
    custom_fields = {
        "Sales Invoice": [
            {
                "fieldname": "si_itemwise_rate_details",
                "label": "Itemwise Rate Details",
                "fieldtype": "Table",
                "options": "Sales Invoice Itemwise Rate Details",
                "insert_after": "items",
                "read_only": 1,
                "depends_on": "eval: doc.docstatus == 0",
            }
        ]
    }
    create_custom_fields(
        custom_fields, ignore_validate=frappe.flags.in_patch, update=update
    )

@frappe.whitelist()
def fetch_rate_details(item_code):
    doc_count = 0
    rate_details = []
    po_details = frappe.get_all('Sales Invoice Item',['rate','parent'],{'item_code':item_code,'parenttype':'Sales Invoice'},order_by="modified")
    for row in po_details[::-1]:
        if frappe.db.get_value('Sales Invoice', row.parent,'docstatus') == 1:
            po_doc = frappe.get_doc('Sales Invoice', row.parent)
            rate_details.append(
                {
                'sales_invoice': row.parent,
                'date': po_doc.posting_date,
                'customer': po_doc.customer, 
                'rate': row.rate}
            )
            doc_count += 1
        if doc_count == 5:
            break
    return rate_details

@frappe.whitelist()
def uom_list(item):
    uom_list=frappe.db.get_list('UOM Conversion Detail',{"parent":item},'uom')
    new_uoms = []
    for uom in uom_list:
        new_uoms.append(uom['uom'])
    return new_uoms



def update_si_to_dn(doc, action):
	for row in doc.items:
		if row.item_code and row.uom:
			uom_list=frappe.db.get_list('UOM Conversion Detail',{"parent":row.item_code},'uom')
			new_uoms = []
			for uom in uom_list:
				new_uoms.append(uom['uom'])
			if row.uom not in new_uoms:
				frappe.throw(_(f"UOM {row.uom} is invalid for the item {row.item_code} in the row {row.idx}"))
		if row.delivery_note:
			frappe.db.set_value('Delivery Note Item', {'parent':row.delivery_note, 'item_code': row.item_code}, 'against_sales_invoice', doc.name)
